web/src/copy_and_paste.ts

```diff
@@ -183,6 +183,72 @@ async function copy_selection_to_clipboard(selection: Selection): Promise<void>
         document.removeEventListener("copy", cb);
     }
 }
+
+// We want to grab the closest katex-display up the tree
+// in cases where we can resolve the selected katex expression
+// from a math block into an inline expression.
+// The returned element from this function
+// is the one we call 'closest' on.
+function get_closest_element(node: Node | null): Element | null {
+    if (node === null || node instanceof Element) {
+        return node;
+    }
+    return node.parentElement;
+}
+
+/*
+    This is done to make the copying within math blocks smarter.
+    We mutate the selection for selecting singular katex expressions
+    within math blocks when applicable, which will be
+    converted into the inline $$<expr>$$ syntax.
+
+    In case when a single expression or its subset is selected
+    within a math block, we adjust the selection so that it
+    selects the katex span which is parenting that expression.
+
+    We do not want to do this if the selection
+    spreads across multiple katex displays or both the
+    focus and anchor are not within the same katex-display.
+    That is why we call `undo_katex_selection_if_applicable`
+    at the end.
+*/
+function improve_katex_selection_range(selection: Selection): void {
+    const original_range = selection.getRangeAt(0);
+    const anchor_element = get_closest_element(selection.anchorNode);
+    const focus_element = get_closest_element(selection.focusNode);
+    if (anchor_element) {
+        const parent = anchor_element.closest(".katex-display");
+        const is_math_block = parent !== null && parent !== selection.anchorNode;
+        if (is_math_block) {
+            const range = document.createRange();
+            range.selectNodeContents(parent);
+            selection.removeAllRanges();
+            selection.addRange(range);
+        }
+    } else if (focus_element) {
+        const parent = focus_element.closest(".katex-display");
+        const is_math_block = parent !== null && parent !== selection.focusNode;
+        if (is_math_block) {
+            const range = document.createRange();
+            range.selectNodeContents(parent);
+            selection.removeAllRanges();
+            selection.addRange(range);
+        }
+    }
+
+    // If the anchor and focus end up in different katex-displays, this isn't
+    // meant to be an inline expression, so we undo what was done previously.
+    if (
+        focus_element &&
+        anchor_element &&
+        focus_element?.closest(".katex-display") !== anchor_element?.closest(".katex-display")
+    ) {
```

Comment:
Could you do this check at the top of the function and just return, rather than needing to undo what was done above? That seems like it might be a lot simpler.